// music-metadata/lib/ParserFactory.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fileType = require("file-type");
const MimeType = require("media-typer");
const _debug = require("debug");
const MetadataCollector_1 = require("./common/MetadataCollector");
const AiffParser_1 = require("./aiff/AiffParser");
const APEv2Parser_1 = require("./apev2/APEv2Parser");
const AsfParser_1 = require("./asf/AsfParser");
const FlacParser_1 = require("./flac/FlacParser");
const MP4Parser_1 = require("./mp4/MP4Parser");
const MpegParser_1 = require("./mpeg/MpegParser");
const musepack_1 = require("./musepack");
const OggParser_1 = require("./ogg/OggParser");
const WaveParser_1 = require("./riff/WaveParser");
const WavPackParser_1 = require("./wavpack/WavPackParser");
const debug = _debug("music-metadata:parser:factory");
class ParserFactory {
    constructor() {
        // ToDo: expose warnings to API
        this.warning = [];
    }
    /**
     *  Parse metadata from tokenizer
     * @param {ITokenizer} tokenizer
     * @param {string} contentType
     * @param {IOptions} opts
     * @returns {Promise<INativeAudioMetadata>}
     */
    static parse(tokenizer, contentType, opts) {
        return __awaiter(this, void 0, void 0, function* () {
            // Resolve parser based on MIME-type or file extension
            let parserId = ParserFactory.getParserIdForMimeType(contentType) || ParserFactory.getParserIdForExtension(contentType);
            if (!parserId) {
                // No MIME-type mapping found
                debug("No parser found for MIME-type / extension: " + contentType);
                const buf = Buffer.alloc(4100);
                yield tokenizer.peekBuffer(buf, 0, buf.byteLength, tokenizer.position, true);
                const guessedType = fileType(buf);
                if (!guessedType)
                    throw new Error("Failed to guess MIME-type");
                parserId = ParserFactory.getParserIdForMimeType(guessedType.mime);
                if (!parserId)
                    throw new Error("Guessed MIME-type not supported: " + guessedType.mime);
                return this._parse(tokenizer, parserId, opts);
            }
            // Parser found, execute parser
            return this._parse(tokenizer, parserId, opts);
        });
    }
    /**
     * @param filePath Path, filename or extension to audio file
     * @return Parser sub-module name
     */
    static getParserIdForExtension(filePath) {
        if (!filePath)
            return;
        const extension = this.getExtension(filePath).toLocaleLowerCase() || filePath;
        switch (extension) {
            case ".mp2":
            case ".mp3":
            case ".m2a":
                return 'mpeg';
            case ".ape":
                return 'apev2';
            case ".aac":
            case ".mp4":
            case ".m4a":
            case ".m4b":
            case ".m4pa":
            case ".m4v":
            case ".m4r":
            case ".3gp":
                return 'mp4';
            case ".wma":
            case ".wmv":
            case ".asf":
                return 'asf';
            case ".flac":
                return 'flac';
            case ".ogg":
            case ".ogv":
            case ".oga":
            case ".ogm":
            case ".ogx":
            case ".opus": // recommended filename extension for Ogg Opus
            case ".spx": // recommended filename extension for Ogg Speex
                return 'ogg';
            case ".aif":
            case ".aiff":
            case ".aifc":
                return 'aiff';
            case ".wav":
                return 'riff';
            case ".wv":
            case ".wvp":
                return 'wavpack';
            case ".mpc":
                return 'musepack';
        }
    }
    static loadParser(moduleName, options) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (moduleName) {
                case 'aiff': return new AiffParser_1.AIFFParser();
                case 'apev2': return new APEv2Parser_1.APEv2Parser();
                case 'asf': return new AsfParser_1.AsfParser();
                case 'flac': return new FlacParser_1.FlacParser();
                case 'mp4': return new MP4Parser_1.MP4Parser();
                case 'mpeg': return new MpegParser_1.MpegParser();
                case 'musepack': return new musepack_1.default();
                case 'ogg': return new OggParser_1.OggParser();
                case 'riff': return new WaveParser_1.WaveParser();
                case 'wavpack': return new WavPackParser_1.WavPackParser();
                default:
                    throw new Error(`Unknown parser type: ${moduleName}`);
            }
        });
    }
    static _parse(tokenizer, parserId, opts = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            // Parser found, execute parser
            const parser = yield ParserFactory.loadParser(parserId, opts);
            const metadata = new MetadataCollector_1.MetadataCollector(opts);
            yield parser.init(metadata, tokenizer, opts).parse();
            return metadata.toCommonMetadata();
        });
    }
    static getExtension(fname) {
        const i = fname.lastIndexOf('.');
        return i === -1 ? '' : fname.slice(i);
    }
    /**
     * @param {string} mimeType MIME-Type, extension, path or filename
     * @returns {string} Parser sub-module name
     */
    static getParserIdForMimeType(mimeType) {
        let mime;
        try {
            mime = MimeType.parse(mimeType);
        }
        catch (err) {
            debug(`Invalid MIME-type: ${mimeType}`);
            return;
        }
        const subType = mime.subtype.indexOf('x-') === 0 ? mime.subtype.substring(2) : mime.subtype;
        switch (mime.type) {
            case 'audio':
                switch (subType) {
                    case 'mp3': // Incorrect MIME-type, Chrome, in Web API File object
                    case 'mpeg':
                        return 'mpeg'; // ToDo: handle ID1 header as well
                    case 'flac':
                        return 'flac';
                    case 'ape':
                    case 'monkeys-audio':
                        return 'apev2';
                    case 'mp4':
                    case 'aac':
                    case 'aacp':
                    case 'm4a':
                        return 'mp4';
                    case 'ogg': // RFC 7845
                    case 'opus': // RFC 6716
                    case 'speex': // RFC 5574
                        return 'ogg';
                    case 'ms-wma':
                    case 'ms-wmv':
                    case 'ms-asf':
                        return 'asf';
                    case 'aiff':
                    case 'aif':
                    case 'aifc':
                        return 'aiff';
                    case 'vnd.wave':
                    case 'wav':
                    case 'wave':
                        return 'riff';
                    case 'wavpack':
                        return 'wavpack';
                    case 'musepack':
                        return 'musepack';
                }
                break;
            case 'video':
                switch (subType) {
                    case 'ms-asf':
                    case 'ms-wmv':
                        return 'asf';
                    case 'm4v':
                    case 'mp4':
                        return 'mp4';
                    case 'ogg':
                        return 'ogg';
                }
                break;
            case 'application':
                switch (subType) {
                    case 'vnd.ms-asf':
                        return 'asf';
                    case 'ogg':
                        return 'ogg';
                }
                break;
        }
    }
}
exports.ParserFactory = ParserFactory;
